Added Set/Get Chef Extension commands

siddheshwar-more opened the pull request:
Implemented following chef extension related commands-
1. azure vm extension set-chef
2. azure vm extension get-chef

set-chef - Adds the chef extension to VM
get-chef - Gives chef extension details  

Thank you!

azurecla commented:
Hi __@siddheshwar-more__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, AZPRBOT;

amarzavery commented on lib/commands/asm/vm._js:
```diff
@@ -177,6 +177,23 @@ exports.init = function(cli) {
+    .option('-O, --validation-pem <validation-pem>', $('chef validation pem file path'))
+    .option('-c, --client-config <client-config>', $('chef client configuration file(i.e client.rb) path'))
+    .option('-b, --auto-update-client', $('auto update chef client'))
+    .option('-b, --disable', $('disable extension'))
```
you cannot have same short hand switch names for two parameters

amarzavery commented on lib/commands/asm/vm/vmclient.js:
```diff
@@ -2743,34 +2844,39 @@ _.extend(VMClient.prototype, {
+  getChefExtension: function(vmName, options, callback, logger) {
+    var self = this;
+
+    options.publisherName = 'Chef.Bootstrap.WindowsAzure';
```
Will the publisher name always be 'Chef.Bootstrap.WindowsAzure' ? If so, how about making it a constant?

amarzavery commented on lib/commands/asm/vm/vmclient.js:
```diff
@@ -2690,38 +2679,150 @@ _.extend(VMClient.prototype, {
+    }
+
+    // Set defaults
+    var publisherName = 'Chef.Bootstrap.WindowsAzure';
```
Making this a constant at the top of the file would be a better idea.

amarzavery commented:
You need to record the tests while running them live. This will create a recordings file in the test\recordings directory. 

Please have a look at this wiki page on how to record tests and which envt. variables you need to set.
https://github.com/Azure/azure-xplat-cli/wiki/Setting-up-environment-variables-for-running-xplat-cli-tests

siddheshwar-more commented:
@amarzavery 

I've set all environment variable required to run tests.
But still some `cli affinity-group` related tests are failing on my local as well as on Travis.

I can see same tests are failed for #1560 as well! Refer - https://travis-ci.org/Azure/azure-xplat-cli/jobs/49051456#L4127 

I guess its the common issue. If you know the solution, could you please share the details?

Actually my changes are related to extension commands. I ran only extension related tests by commenting other tests from ** tests/testlist.txt** file and all extension tests are passed.

Thank you.

yugangw-msft commented:
I still don't see the recording files. When you run live tests in record mode then you will see recording files in the test/recordings directory. 
You have to check them in as well as a part of the PR.

Do not worry about the tests that are not a part of your work

amarzavery commented:
Some CLI Tests will fail for now as we have had a breaking change in Node SDK and have not yet published a new version of node sdk where those changes have been fixed.

amarzavery commented:
```
set AZURE_NOCK_RECORD=true
```
Comment out other tests in testlist.txt
Only run your tests in Record Mode.
You will see recorded files in test\recordings directory.
The recorded files willl have the same name as ur test files.

To verify that recording was successfull

set envt variable 
```
set AZURE_NOCK_RECORD=
set NOCK_OFF=
```

and run your recorded tests in mock mode..
Your tests should run successfully..

Everything looks good then commit those files in to the PR.

siddheshwar-more commented:
@amarzavery
I've set-up all test environment and also commented other tests. But I got following response during set chef extension unit tests run.

https://gist.github.com/siddheshwar-more/8d6eb4eccec0115fe3f1 

I've compare this ^ file with existing extension recording files. In above 
https://gist.github.com/siddheshwar-more/8d6eb4eccec0115fe3f1#file-cli-vm-extension_set-chef-nock-js-L27 is empty?

Can you please help us in this matter?

Thank you!

amarzavery commented on test/data/set-chef-extension-validation.pem:
```diff
@@ -0,0 +1,3 @@
+-----BEGIN CERTIFICATE-----
+MIIEoQIBAAKCAQEAtV7R+MaM4PFp3I1xVnHE9kgpgZ38qo+2SeXD0R1ifIFNlj3NghOwvgbFicCsoteWZ9Tmxvi1c8IM+H9SIIjcSbycz9Y+XENC7CyP+/rApFxJ0/CRHdT9eA9X9yBhnseAqzTdHC0qY9re+LMWTZe3C7CNwWRqeF+kmaJWe3qfSWAvMEa4xMiGM7Y0B6gzz73jAlce3EkDKJQTEy9fPCKtxbItY1JGzTkoS74PxKI7EoedjpAc4hbKr4n2eqDiuEUtWT6HFlnp63H3vviG3ctsppozRhJF4FzB938c9XwUizvP+PWd7EZh3zWEaAsNtlNL9kXNotF+YvlPNmFiD5+iYQIDAQABAoIBAFK/hqPujoLokehu3elXMco9pTY+avM ...
```
Is this a dummy cert or an actual cert?

siddheshwar-more replied:
Its a dummy cert !

amarzavery commented:
@siddheshwar-more 

I am sure you would have done this, however I want to verify:

You should only set the envt variable for nock recording:
```
set AZURE_NOCK_RECORD=true
```

The envt variable to run tests live should be unset while recording tests:
```
set NOCK_OFF=
```

If this is not done, it may interfere with recording tests.

amarzavery commented on test/commands/cli.vm.extension_get-chef-tests.js:
```diff
@@ -0,0 +1,109 @@
+          result.exitStatus.should.equal(0);
+          var vmList = JSON.parse(result.text);
+          var found = vmList.some(function(vm) {
+            if (vm.OSDisk.operatingSystem.toLowerCase() === 'windows') {
```
What if the subscription does not have any provisioned windows vm ? The test should be robust in creating one if it cannot find one and delete the vm after the test is over.

amarzavery commented:
I was able to record the get-chef tests successfully. I am having a problem running the set-chef tests.

I get the error:
 cli vm extension: Set Chef extensions: error: A VM with dns prefix "xplattestvm" already exists

Reason being there is no clean up (deleting the vm ) done in after each. 

Either do the clean up in after each or make the before each smarter. Create a vm with the name xplattestvm if you cannot fin one. Otherwise use the existing one.


For recording tests:
1. Login using an org id or microsoft.com account (azure login -u$username -p $password)
2. Set the azure subscription under which the testing would be done (azure account set "Subscription name"
3. set NOCK_OFF=
4. set AZURE_NOCK_RECORD=true
5. Create recording files in test/recordings directory if they do not exist
6. Comment out other tests and run your tests in testlist.txt one by one to record.
7. Once tests are recorded, run them mocked. For that set the envt. variables:
set NOCK_OFF=
set AZURE_NOCK_RECORD=

siddheshwar-more commented:
@amarzavery -
Thank you. Now I'm able to record live traffic , working on review comments!

Live traffic has sensitive data related to azure account, I'm planning to replace that with existing extension related test like- https://github.com/Azure/azure-xplat-cli/blob/master/test/recordings/cli.vm.extension-tests.nock.js#L9 
is it fine?

amarzavery commented:
@siddheshwar-more 
That should be fine. The subscription id should not be a problem. It is just a guid. Making sure that recorded file does not have cert or access token is a good idea.

siddheshwar-more commented:
@amarzavery -
I've fixed all review comments related to tests. Also added test recording files and did testing with live and mock mode. All tests are passed.

Could you please merge this PR?

Thank you!

amarzavery commented:
Awesome! This looks good and the CI is passing as well. Good job :+1:
